### DISCUSSION TITLE
Cannot delete role - "Class name must be a valid object or a string" error response

### DISCUSSION THREAD

**Discussion by Raptor-PL** [spatie/laravel-permission#2736] [Q&A]
Hi Guys,

I'm getting this error while I try to delete a role. Getting roles and creating work fine.

My setup:
- Laravel 11 with Sanctum and Spatie Permissions as backend API
- Vue 3 with Axios for the frontend

I only started the project few weeks ago. Laravel installed from scratch, installed API packages. Permissions installed from scratch following docs. Made no changes to anything except those from docs as in previous versions of Laravel everything worked fine...

Yes, there's a similar post about deleting permissions, but it didn't help in my case.

RoleController:
```php
?php

namespace App\Http\Controllers;

use Spatie\Permission\Models\Role;
use Illuminate\Http\Request;
use Illuminate\Validation\Rule;

class RoleController extends Controller
{
    /**
     * Display a listing of the resource.
     */
    public function index()
    {
        return Role::with('permissions')->where('id', '>', 1)->get();
    }

    /**
     * Store a newly created resource in storage.
     */
    public function store(Request $request)
    {
        $request->validate([
            'name' => 'required|string|max:255|min:5|unique:roles',
            'description' => 'nullable|string',
            'permissions' => 'nullable|array'
        ]);

        $role = Role::create($request->only(['name', 'description']));

        $permissionsArray = $request->only(['permissions']);

        if($role && $permissionsArray) {
            $role->syncPermissions($permissionsArray);
        }
    }

    /**
     * Display the specified resource.
     */
    public function show(string $id)
    {
        return Role::with('permissions')->findOrFail($id);
    }

    /**
     * Update the specified resource in storage.
     */
    public function update(Request $request, string $id)
    {
        
        $request->validate([
            'name' => ['required', 'string', 'max:255', 'min:5', Rule::unique('roles')->ignore($id)],
            'description' => 'nullable|string',
            'permissions' => 'nullable|array'
        ]);
        
        $role = Role::findOrFail($id);        

        $role->update($request->only(['name','description']));

        $permissionsArray = $request->only(['permissions']);
        
        if($permissionsArray) {
            $role->syncPermissions($permissionsArray);
        }
    }

    /**
     * Remove the specified resource from storage.
     */
    public function destroy(string $id)
    {
        if($id > 3) {
            
            $role = Role::findById($id);

            $role->delete();

            return response()->json(['message'=>'Resource deleted.'], 204);
        } else {
            return response()->json([
                'errors'=> 'This resource cannot be deleted.'
            ], 500);
        }
    }
}
``` 
![image](https://github.com/user-attachments/assets/bf3fca01-1f7a-4d88-9a25-4b65bef92676)

**Comment by drbyte** [comment]
`Class name must be a valid object or string` is a PHP error triggered by calling `new` on an unknown class.

Given that you didn't mention your actual current Laravel version, I'm going to assume you meant today's latest, which (for me right now) is `11.28.0`. In that case, line `855` is inside `function newRelatedInstance($class)` and running `tap(new $class, ...)` ... presumably to query records related to the `Role` object.
So, I'm guessing that (in order to cascade the delete onto related models) it's having trouble resolving a relation, since otherwise it shouldn't have trouble instantiating the class.

1. If this is part of an upgrade, first clear your artisan `config` cache (`php artisan config:clear`), just in case the cache is broken/stale and not correctly referring to proper models as defined in the config files. (You could also `cache:clear` for thoroughness.)
2. Have you added any custom `morphMap()` (ie: `Relation::enforceMorphMap([...])`) to your application? Do those relate to roles/permissions or any users having any assigned roles/permissions?
3. In your database's `model_has_roles` table, what's the `model_type` value on the role records you're trying to delete? How does that relate to the actual models in your application? (example: if you changed namespaces of models, the stored relation names would be outdated)
4. What are your `config/permission.php` settings for all the pivot keys under `column_names`? If they're not the defaults, then debugging will require incorporating whatever those are set to. Actually, have you customized anything in `permission.php`?
